### PR TITLE
Order new ssl cert nsd 11692

### PR DESCRIPTION
**Pull request by calevans**

### CODE REVIEW

File: tests/Resource/Ssl/EndpointTest.php [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+};
+
+class EndpointTest extends EndpointTestCase {
+
+  /** {@inheritDoc} */
+  protected const _RESOURCE_PATH = __DIR__ . '/resources';
+
+  /** {@inheritDoc} */
+  protected const _RESOURCE_GET = 'GET-%2Fssl-cert%2F1.json';
+// what is returned by API ssl-cert/CERT_ID
```
Comment (adrian-enspired) [inline tests/Resource/Ssl/EndpointTest.php]: i don't know

File: src/Resource/Ssl/Ssl.php [hunk 2]
```diff
@@ -23,7 +23,10 @@ class Ssl extends Model {
   public const MODULE_NAME = 'Ssl';
 
   /** {@inheritDoc} */
-  protected const _PROPERTY_ALIASES = ['id' => 'cert_id'];
+  protected const _PROPERTY_ALIASES = [
+    'id' => 'cert_id',
+    'service_id' => 'service.id'
```
Comment (adrian-enspired) [inline src/Resource/Ssl/Ssl.php]: Please confirm whether `service_id` is actually needed. I suspect it was an aberration from problems encountered during troubleshooting tests yesterday. I've ran unit tests and also made requests against demo2 without this param; everything seems to work fine.

File: tests/Resource/Ssl/resources/ssl-cert-1.toArray.json [hunk 3]
```diff
@@ -2,21 +2,21 @@
   "id": 637,
   "alt_domains": "",
   "alt_names": [
-    "www.example.com"
+    "www.voicesoftheelephpant.com"
```
Comment (adrian-enspired) [inline tests/Resource/Ssl/resources/ssl-cert-1.toArray.json]: were these changes intentional?

File: src/Resource/Ssl/Endpoint.php [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+
+  /**
+   * Retrieve a certificate by it's service_id
+   *
+   * @param int $service_id a valid service_id for a certificate
+   * @return Ssl
+   */
+  public function retrieveByServiceId(int $service_id) : Ssl {
+    $filter = ['filter' => ['service_id' => $service_id]];
+    $response = $this->_client->request(
```
[the remark this answers is not in the document]
Comment (calevans) [reply src/Resource/Ssl/Endpoint.php]: yes, certs are 1:1 with service.
done.

File: src/Resource/Ssl/Endpoint.php [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+   * Decode an existing CSR and compare it to the package_id
+   *
+   * @param string $csr The CSR to decode
+   * @param int $package_id valid package_id for the type of csr
+   *
+   * @return array
+   * @throws \GuzzleHttp\Exception\ClientException If request fails
+   */
+  public function decodeCsr(string $csr, int $package_id) : array {
+    return Util::decodeResponse($this->_client->post(
```
[the remark this answers is not in the document]
Comment (calevans) [reply src/Resource/Ssl/Endpoint.php]: fixed 1, left 2

File: src/Resource/Ssl/Ssl.php [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+  /** {@inheritDoc} */
+  protected const _PROPERTY_COLLAPSED = [];
+
+  /** {@inheritDoc} */
+  protected const _PROPERTY_MODELS = [
+    'client' => Client::class
+  ];
+
+  /** {@inheritDoc} */
+  protected const _PROPERTY_NAMES = [
```
[the remark this answers is not in the document]
Comment (calevans) [reply src/Resource/Ssl/Ssl.php]: done

File: tests/Resource/Ssl/resources/ssl-cert-1.toArray.json (hunk 3 above)
Comment (calevans) [reply tests/Resource/Ssl/resources/ssl-cert-1.toArray.json]: that file is gone anyhow.

File: src/Resource/Ssl/SslException.php [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+declare(strict_types  = 1);
+
+namespace Nexcess\Sdk\Resource\Order;
+
+use Nexcess\Sdk\Exception;
+
+/**
+ * Error conditions for order resources.
+ */
+class SslException extends Exception {
```
[the remark this answers is not in the document]
Comment (calevans) [reply src/Resource/Ssl/SslException.php]: yes. it is used and has been fixed.

File: src/Resource/Ssl/Endpoint.php [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+      ]
+    );
+
+    return $this->retrieveByServiceId(
+      Util::decodeResponse($response)['service_id']
+    );
+  }
+
+  /**
+   * Decode an existing CSR and compare it to the package_id
```
[the remark this answers is not in the document]
Comment (calevans) [reply src/Resource/Ssl/Endpoint.php]: the underlying nocworx code compares it, not us.

File: src/Resource/Ssl/Endpoint.php [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+use Nexcess\Sdk\ {
+  Resource\Endpoint as ReadableEndpoint,
+  Resource\Ssl\Ssl,
+  Util\Util
+};
+
+/**
+ * API endpoint for Ssl.
+ */
+class Endpoint extends ReadableEndpoint {
```
[the remark this answers is not in the document]
Comment (calevans) [reply src/Resource/Ssl/Endpoint.php]: no

File: src/Resource/Ssl/Ssl.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (adrian-enspired) [reply src/Resource/Ssl/Ssl.php]: still some overlap here (client_id). should probably go in readonly.

question: is there _anything_ here which is editable? not including via actions like renewing the service, which would change the expiry.

if not, then realistically everything should be in READONLY, except for the cert_id.

File: src/Resource/Ssl/Ssl.php [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+      );
+  }
+
+  /**
+   * Decode an existing CSR and compare it to the package_id
+   *
+   * @return array
+   * @throws \GuzzleHttp\Exception\ClientException on fail
+   */
+  public function decodeCsr() : array {
```
Comment (adrian-enspired) [inline src/Resource/Ssl/Ssl.php]: should this (and `getCsrDetails` below) be proxied?  seems to me like it would make sense for them to only exist on the endpoint.  if there's a reason they should be proxied, that's fine.

File: src/Resource/Ssl/Ssl.php [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+   * and key and then create the certificate.
+   *
+   * @return Ssl
+   * @throws \GuzzleHttp\Exception\ClientException on fail
+   */
+  public function create() : Ssl {
+    $endpoint = $this->_getEndpoint();
+    assert($endpoint instanceof Endpoint);
+
+    if (!empty($this->get('csr'))) {
```
Comment (adrian-enspired) [inline src/Resource/Ssl/Ssl.php]: `csr` is checked, but no other values are. If they're missing, we'll get a `TypeError`, which is unexpected when invoking a method with no arguments. Should we check first and throw?

File: src/Resource/Ssl/Endpoint.php [hunk 12]
```diff
@@ -0,0 +1,244 @@
...
+   * @param int $months The number of months to make this certificate valid for.
+   * @param int $package_id The SSL package purchased
+   * @param array $approver_email format
+   *              'domain.name' => 'approver@domain.name' Must be one of the
+   *              approved 'approver emails'
+   *
+   * @return Ssl
+   * @throws \GuzzleHttp\Exception\ClientException If request fails
+   */
+  public function createCertificateFromCsr(
```
Comment (adrian-enspired) [inline src/Resource/Ssl/Endpoint.php]: should rename `createFromCsr`

File: src/Util/lang/en_US.json [hunk 13]
```diff
@@ -110,6 +110,42 @@
...
+      "chain":"An optional chain certificate"
+    },
+    "createCertificateFromCsr" : {
+      "csr" : "The Certificate Signing Request",
+      "key" : "The private key for this certificate",
+      "months" : "How many months this certificate is good for",
+      "package_id" : "The NocWorx package id appropriate for this type of certificate.",
+      "approver_email" :"Array of approver email addresses, one per domain. Must be from the list of approver email addresses!"
+    },
+    "createCertificate" : {
```
Comment (adrian-enspired) [inline src/Util/lang/en_US.json]: needs to be renamed `create` to match method name

File: src/Util/lang/en_US.json [hunk 14]
```diff
@@ -110,6 +110,42 @@
...
+  "Ssl": {
+    "retrieveByServiceId" : {
+      "service_id" : "The service id for the certificate you wish to retrieve."
+    },
+    "importCertificate" : {
+      "key": "The private key for this certificate",
+      "crt":"The certificate",
+      "chain":"An optional chain certificate"
+    },
+    "createCertificateFromCsr" : {
```
Comment (adrian-enspired) [inline src/Util/lang/en_US.json]: `createFromCsr`

File: src/Resource/Ssl/Ssl.php (hunk 10 above)
Comment (calevans) [reply src/Resource/Ssl/Ssl.php]: I proxied them by default but yes, in hindsight, it only makes sense on the Endpoint. POSSIBLY a static method on Ssl?

File: src/Resource/Ssl/Ssl.php (hunk 11 above)
Comment (calevans) [reply src/Resource/Ssl/Ssl.php]: not in favor of it, no. We already validate the parameters at the endpoint. That having been said, if you feel strongly about it, we can.

File: src/Resource/Ssl/Ssl.php (hunk 10 above)
Comment (calevans) [reply src/Resource/Ssl/Ssl.php]: removed

File: src/Resource/Ssl/Ssl.php (hunk 11 above)
Comment (calevans) [reply src/Resource/Ssl/Ssl.php]: the check for csr is to determine which to create. I don't really care about the CSR itself, it's just a signal at this point. It could contain HELLO WORLD and we'd try to use it.

File: src/Resource/Ssl/Ssl.php (hunk 6 above)
[the remark this answers is not in the document]
Comment (calevans) [reply src/Resource/Ssl/Ssl.php]: right now we ONLY support creating certs/ We do not support editing existing certs. So, I would guess, no, there is nothing EDITABLE. However, I was unclear if READONLY really meant READONLY or if it mean that we don't send it to the API, we only get it back.
